src/java.base/share/classes/jdk/internal/foreign/layout/AbstractGroupLayout.java

```diff
@@ -132,6 +140,9 @@ enum Kind {
         long sizeof(List<MemoryLayout> elems) {
             long size = 0;
             for (MemoryLayout elem : elems) {
+                if (this == STRUCT && (size % elem.bitAlignment() != 0)) {
```

Comment:
We might drop `sizeOp` and instead have two different `sizeof` implementations in a more OO style.

Reply:
I've did another pass on the impl, where I've greatly simplified AbstractGroupLayoutImpl, StructLayoutImpl and UnionLayoutImpl. Now all the computation occurs in their factories (which is better, as the stack trace in case of an issue is more readable), rather than inside the enum methods. Also, the new classes avoid any kind of re-computation: for each group, we compute size and alignment upfront. We then store the original alignment into a new field (`minBitAlignment`) so that we can use that to determine validity of subsequent `withBitAlignment` operations.